runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java

```diff
@@ -0,0 +1,171 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.apex;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.apache.beam.runners.apex.translators.TranslationContext;
+import org.apache.beam.runners.core.UnboundedReadFromBoundedSource;
+import org.apache.beam.runners.core.UnboundedReadFromBoundedSource.BoundedToUnboundedSourceAdapter;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.Read;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.transforms.windowing.WindowFn;
+import org.apache.beam.sdk.util.AssignWindows;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.api.Context.DAGContext;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.LocalMode;
+import com.datatorrent.api.StreamingApplication;
+import com.google.common.base.Throwables;
+
+/**
+ * A {@link PipelineRunner} that translates the
+ * pipeline to an Apex DAG and executes it on an Apex cluster.
+ * <p>
+ * Currently execution is always in embedded mode,
+ * launch on Hadoop cluster will be added in subsequent iteration.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ApexRunner extends PipelineRunner<ApexRunnerResult> {
+
+  private final ApexPipelineOptions options;
+
+  public ApexRunner(ApexPipelineOptions options) {
+    this.options = options;
+  }
+
+  public static ApexRunner fromOptions(PipelineOptions options) {
+    return new ApexRunner((ApexPipelineOptions) options);
+  }
+
+  @Override
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+      PTransform<InputT, OutputT> transform, InputT input) {
+    if (Window.Bound.class.equals(transform.getClass())) {
+      return (OutputT) ((PCollection) input).apply(
+          new AssignWindowsAndSetStrategy((Window.Bound) transform));
+    } else if (Create.Values.class.equals(transform.getClass())) {
+      return (OutputT) PCollection
+          .<OutputT>createPrimitiveOutputInternal(
+              input.getPipeline(),
+              WindowingStrategy.globalDefault(),
+              PCollection.IsBounded.BOUNDED);
+    } else if (Read.Bounded.class.equals(transform.getClass())) {
+      return (OutputT) ((PBegin) input).apply(new UnboundedReadFromBoundedSource<>(((Read.Bounded)transform).getSource()));
```

Comment:
I believe this line may be the cause of your trouble with `CountingInputTest`, or at least the `testBoundedInput` method. Since this converts a bounded `PCollection` to an unbounded `PCollection`, it changes what graphs can be successfully constructed.

The way of the future here is to not use `apply` to do the overrides, but to do them after pipeline construction is complete, during translation. The Flink and Spark runners have examples of doing this in the traversal - we intend to refactor that logic out so that replacements are trivial.